app/lib/package/export_api_to_cdn.dart

```diff
@@ -0,0 +1,239 @@
+import 'package:collection/collection.dart';
+import 'package:crypto/crypto.dart';
+import 'package:gcloud/storage.dart';
+import 'package:logging/logging.dart';
+import 'package:meta/meta.dart';
+import 'package:pool/pool.dart';
+import 'package:retry/retry.dart';
+
+import '../shared/datastore.dart';
+import '../shared/storage.dart';
+import '../shared/utils.dart';
+import '../shared/versions.dart';
+import '../task/global_lock.dart';
+import 'backend.dart';
+import 'models.dart';
+
+final Logger _logger = Logger('export_api_to_cdn');
+
+/// The default concurrency to upload API JSON files to the CDN bucket.
+const _defaultUploadPackageApisToCdnBucketConcurrency = 8;
+
+List<String> _cdnObjectNames(String package) => [
+      '$runtimeVersion/api/packages/$package',
+      'current/api/packages/$package',
+    ];
+
+/// Runs a forever loop and tries to get a global lock.
+///
+/// Once it has the claim, it scans the packages entities and uploads
+/// the package API JSONs to the CDN bucket.
+/// Tracks the package updates for the next up-to 24 hours and writes
+/// the API JSONs after every 5 minutes.
+///
+/// When other process has the claim, the loop waits a minute before
+/// attempting to get the claim.
+Future<Never> uploadPackageApisToCdnBucketInForeverLoop(Bucket bucket) async {
+  final lock = GlobalLock.create(
+    '$runtimeVersion/package/update-api-cdn',
+    expiration: Duration(minutes: 20),
+  );
+  while (true) {
+    try {
+      await lock.withClaim((claim) async {
+        await doUploadPackageApisToCdnBucket(claim, bucket);
+      });
+    } catch (e, st) {
+      _logger.warning('Package API CDN update failed.', e, st);
+    }
+    // Wait for 1 minutes for sanity, before trying again.
+    await Future.delayed(Duration(minutes: 1));
+  }
+}
+
+@visibleForTesting
+Future<void> doUploadPackageApisToCdnBucket(
+  GlobalLockClaim claim,
+  Bucket bucket, {
+  Duration sleepDuration = const Duration(minutes: 2),
+  int concurrency = _defaultUploadPackageApisToCdnBucketConcurrency,
+}) async {
+  final firstClaimed = clock.now();
+
+  // The claim will be released after a day, another process may
+  // start to upload the API JSONs from scratch again.
+  final workUntil = clock.now().add(Duration(days: 1));
+
+  Future<void> update(String package, bool isVisible) async {
+    if (isVisible) {
+      await _uploadPackageApiToCdnBucket(package, bucket);
+    } else {
+      await _deletesPackageFromBucket(package, bucket);
+    }
+  }
+
+  // initial scan of packages
+  final pool = Pool(concurrency);
+  final futures = <Future>[];
+  await for (final mp in dbService.query<ModeratedPackage>().run()) {
+    if (!claim.valid) {
+      return;
+    }
+    final f = pool.withResource(() => update(mp.name!, false));
+    futures.add(f);
+  }
+  await Future.wait(futures);
+  futures.clear();
+  if (!claim.valid) {
+    return;
+  }
+
+  await for (final package in dbService.query<Package>().run()) {
```

Comment:
Will a package ever be in both ModeratedPackage and Package at the same time?

Reply:
It won't at the same time, at least in the long run. But as the queries happen with some delay, a package may appear in both for a short time.